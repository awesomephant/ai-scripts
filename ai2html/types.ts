export type ImageFormat = "auto" | "png" | "png24" | "jpg" | "svg"
export type ResponsivenessOption = "fixed" | "dynamic"
type TextRenderingMethod = "html" | "image"

export interface FontRule {
	aifont: string
	family: string
	weight: string
	style: string
}

export interface ArtboardGroupForOutput {
	groupName: string
	artboards: Artboard[]
}

export type SettingsTextBlockMode =
	| "css"
	| "js"
	| "html"
	| "settings"
	| "text"
	| "html-before"
	| "html-after"

export interface ai2HTMLSettings {
	namespace: string
	settings_version: string
	create_promo_image: boolean
	promo_image_width: number
	image_format: ImageFormat[]
	write_image_files: boolean
	responsiveness: ResponsivenessOption
	text_responsiveness: ResponsivenessOption
	/**
	 * Inserts a max-width css instruction on the div containing the ai2html partial. The width should be specified in pixels.
	 */
	max_width: number
	output:
		| "one-file"
		| "multiple-files"
		| "one-file-for-all-artboards" // @deprecated
		| "one-file-for-all-artboards" // @deprecated
		| "preview-one-file" // @deprecated
		| "one-file-per-artboard" // @deprecated
		| "preview-multiple-files" // @deprecated
	/**
	 * Defaults to the name of the AI file
	 */
	project_name?: string
	html_output_path: string
	html_output_extension: string
	image_output_path: string
	image_source_path: string
	image_alt_text: string
	alt_text: string
	cache_bust_token?: number
	create_config_file: boolean
	scriptVersion: string
	create_json_config_files: boolean
	image_width: number
	grouped_artboards: boolean
	/**
	 * Create a text block in the AI doc with common settings
	 */
	create_settings_block: boolean
	config_file_path: string
	local_preview_template: string
	png_transparent: boolean
	png_number_of_colors: number // 1-256
	jpg_quality: number // 1-100
	center_html_output: boolean
	use_2x_images_if_possible: boolean
	use_lazy_loader: boolean
	include_resizer_widths: boolean
	fonts?: FontRule[]
	/**
	 * container-query resizing
	 */
	include_resizer_css: boolean
	include_resizer_script: boolean
	/**
	 * Embed background image SVG in HTML instead of loading a file
	 */
	inline_svg: boolean
	/**
	 * Prefix SVG ids with a string to disambiguate from other ids on the page
	 */
	svg_id_prefix: string
	svg_embed_images: boolean
	render_text_as: TextRenderingMethod
	render_rotated_skewed_text_as: TextRenderingMethod
	/**
	 * Render text in both bg image and HTML to test HTML text placement
	 */
	testing_mode: boolean // TODO: bad name
	show_completion_dialog_box: boolean
	/**
	 * Add a URL to make the entire graphic a clickable link
	 */
	clickable_link: string // TODO bad name
	last_updated_text: string
	headline: string
	leadin: string
	summary: string
	notes: string
	sources: string
	credit: string
	/**
	 * List of settings to include in the "ai2html-settings" text block
	 */
	settings_block: (keyof ai2HTMLSettings)[]
	config_file: (keyof ai2HTMLSettings)[]

	/**
	 * NYT-specific
	 */
	project_type: string
	dark_mode_compatible: boolean
	show_in_compatible_apps: boolean
}
