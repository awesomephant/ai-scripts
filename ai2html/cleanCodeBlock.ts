import {
	stripTag,
	straightenCurlyQuotes,
	addEnclosingTag,
	cleanHtmlText,
	straightenCurlyQuotesInsideAngleBrackets
} from "../common/stringUtils"

import { SettingsTextBlockMode } from "./types"

// Clean the contents of custom JS, CSS and HTML blocks
// (e.g. undo Illustrator's automatic quote conversion, where applicable)

/**
 *
 * @param mode
 * @param raw
 * @returns
 */
export default function cleanCodeBlock(mode: SettingsTextBlockMode, raw: string) {
	if (mode.indexOf("html") >= 0) {
		return cleanHtmlText(straightenCurlyQuotesInsideAngleBrackets(raw))
	} else if (mode === "js") {
		// TODO: consider preserving curly quotes inside quoted strings
		return addEnclosingTag("script", straightenCurlyQuotes(raw))
	} else if (mode === "css") {
		return stripTag("style", straightenCurlyQuotes(raw))
	}
	return raw
}
