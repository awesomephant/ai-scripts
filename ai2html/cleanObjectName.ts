import { makeKeyword, trim } from "../common/stringUtils"

/**
 * Remove any annotations and colon separator from an object name
 * @param name string
 * @returns string
 */
export default function cleanObjectName(name: string): string {
	return makeKeyword(name.replace(/^(.+):.*$/, "$1"))
}
