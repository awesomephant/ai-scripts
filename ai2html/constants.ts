import { ai2HTMLSettings, FontRule } from "./types"

const defaultFonts: FontRule[] = [
	{
		aifont: "ArialMT",
		family: "arial,helvetica,sans-serif",
		weight: "",
		style: ""
	},
	{
		aifont: "Arial-BoldMT",
		family: "arial,helvetica,sans-serif",
		weight: "bold",
		style: ""
	},
	{
		aifont: "Arial-ItalicMT",
		family: "arial,helvetica,sans-serif",
		weight: "",
		style: "italic"
	},
	{
		aifont: "Arial-BoldItalicMT",
		family: "arial,helvetica,sans-serif",
		weight: "bold",
		style: "italic"
	},
	{
		aifont: "Georgia",
		family: "georgia,'times new roman',times,serif",
		weight: "",
		style: ""
	},
	{
		aifont: "Georgia-Bold",
		family: "georgia,'times new roman',times,serif",
		weight: "bold",
		style: ""
	},
	{
		aifont: "Georgia-Italic",
		family: "georgia,'times new roman',times,serif",
		weight: "",
		style: "italic"
	},
	{
		aifont: "Georgia-BoldItalic",
		family: "georgia,'times new roman',times,serif",
		weight: "bold",
		style: "italic"
	}
]

// html entity substitution
// TODO: make this a hash map
const basicCharacterReplacements = [
	["\x26", "&amp;"],
	["\x22", "&quot;"],
	["\x3C", "&lt;"],
	["\x3E", "&gt;"]
]

const extraCharacterReplacements = [
	["\xA0", "&nbsp;"],
	["\xA1", "&iexcl;"],
	["\xA2", "&cent;"],
	["\xA3", "&pound;"],
	["\xA4", "&curren;"],
	["\xA5", "&yen;"],
	["\xA6", "&brvbar;"],
	["\xA7", "&sect;"],
	["\xA8", "&uml;"],
	["\xA9", "&copy;"],
	["\xAA", "&ordf;"],
	["\xAB", "&laquo;"],
	["\xAC", "&not;"],
	["\xAD", "&shy;"],
	["\xAE", "&reg;"],
	["\xAF", "&macr;"],
	["\xB0", "&deg;"],
	["\xB1", "&plusmn;"],
	["\xB2", "&sup2;"],
	["\xB3", "&sup3;"],
	["\xB4", "&acute;"],
	["\xB5", "&micro;"],
	["\xB6", "&para;"],
	["\xB7", "&middot;"],
	["\xB8", "&cedil;"],
	["\xB9", "&sup1;"],
	["\xBA", "&ordm;"],
	["\xBB", "&raquo;"],
	["\xBC", "&frac14;"],
	["\xBD", "&frac12;"],
	["\xBE", "&frac34;"],
	["\xBF", "&iquest;"],
	["\xD7", "&times;"],
	["\xF7", "&divide;"],
	["\u0192", "&fnof;"],
	["\u02C6", "&circ;"],
	["\u02DC", "&tilde;"],
	["\u2002", "&ensp;"],
	["\u2003", "&emsp;"],
	["\u2009", "&thinsp;"],
	["\u200C", "&zwnj;"],
	["\u200D", "&zwj;"],
	["\u200E", "&lrm;"],
	["\u200F", "&rlm;"],
	["\u2013", "&ndash;"],
	["\u2014", "&mdash;"],
	["\u2018", "&lsquo;"],
	["\u2019", "&rsquo;"],
	["\u201A", "&sbquo;"],
	["\u201C", "&ldquo;"],
	["\u201D", "&rdquo;"],
	["\u201E", "&bdquo;"],
	["\u2020", "&dagger;"],
	["\u2021", "&Dagger;"],
	["\u2022", "&bull;"],
	["\u2026", "&hellip;"],
	["\u2030", "&permil;"],
	["\u2032", "&prime;"],
	["\u2033", "&Prime;"],
	["\u2039", "&lsaquo;"],
	["\u203A", "&rsaquo;"],
	["\u203E", "&oline;"],
	["\u2044", "&frasl;"],
	["\u20AC", "&euro;"],
	["\u2111", "&image;"],
	["\u2113", ""],
	["\u2116", ""],
	["\u2118", "&weierp;"],
	["\u211C", "&real;"],
	["\u2122", "&trade;"],
	["\u2135", "&alefsym;"],
	["\u2190", "&larr;"],
	["\u2191", "&uarr;"],
	["\u2192", "&rarr;"],
	["\u2193", "&darr;"],
	["\u2194", "&harr;"],
	["\u21B5", "&crarr;"],
	["\u21D0", "&lArr;"],
	["\u21D1", "&uArr;"],
	["\u21D2", "&rArr;"],
	["\u21D3", "&dArr;"],
	["\u21D4", "&hArr;"],
	["\u2200", "&forall;"],
	["\u2202", "&part;"],
	["\u2203", "&exist;"],
	["\u2205", "&empty;"],
	["\u2207", "&nabla;"],
	["\u2208", "&isin;"],
	["\u2209", "&notin;"],
	["\u220B", "&ni;"],
	["\u220F", "&prod;"],
	["\u2211", "&sum;"],
	["\u2212", "&minus;"],
	["\u2217", "&lowast;"],
	["\u221A", "&radic;"],
	["\u221D", "&prop;"],
	["\u221E", "&infin;"],
	["\u2220", "&ang;"],
	["\u2227", "&and;"],
	["\u2228", "&or;"],
	["\u2229", "&cap;"],
	["\u222A", "&cup;"],
	["\u222B", "&int;"],
	["\u2234", "&there4;"],
	["\u223C", "&sim;"],
	["\u2245", "&cong;"],
	["\u2248", "&asymp;"],
	["\u2260", "&ne;"],
	["\u2261", "&equiv;"],
	["\u2264", "&le;"],
	["\u2265", "&ge;"],
	["\u2282", "&sub;"],
	["\u2283", "&sup;"],
	["\u2284", "&nsub;"],
	["\u2286", "&sube;"],
	["\u2287", "&supe;"],
	["\u2295", "&oplus;"],
	["\u2297", "&otimes;"],
	["\u22A5", "&perp;"],
	["\u22C5", "&sdot;"],
	["\u2308", "&lceil;"],
	["\u2309", "&rceil;"],
	["\u230A", "&lfloor;"],
	["\u230B", "&rfloor;"],
	["\u2329", "&lang;"],
	["\u232A", "&rang;"],
	["\u25CA", "&loz;"],
	["\u2660", "&spades;"],
	["\u2663", "&clubs;"],
	["\u2665", "&hearts;"],
	["\u2666", "&diams;"]
]

// CSS text-transform equivalents
const caps = [
	{ ai: "FontCapsOption.NORMALCAPS", html: "none" },
	{ ai: "FontCapsOption.ALLCAPS", html: "uppercase" },
	{ ai: "FontCapsOption.SMALLCAPS", html: "uppercase" }
]

// CSS text-align equivalents
const align = [
	{ ai: "Justification.LEFT", html: "left" },
	{ ai: "Justification.RIGHT", html: "right" },
	{ ai: "Justification.CENTER", html: "center" },
	{ ai: "Justification.FULLJUSTIFY", html: "justify" },
	{ ai: "Justification.FULLJUSTIFYLASTLINELEFT", html: "justify" },
	{ ai: "Justification.FULLJUSTIFYLASTLINECENTER", html: "justify" },
	{ ai: "Justification.FULLJUSTIFYLASTLINERIGHT", html: "justify" }
]

const blendModes = [{ ai: "BlendModes.MULTIPLY", html: "multiply" }]

// list of CSS properties used for translating AI text styles
// (used for creating a unique identifier for each style)

const cssTextStyleProperties = [
	//'top' // used with vshift; not independent of other properties
	"position",
	"font-family",
	"font-size",
	"font-weight",
	"font-style",
	"color",
	"line-height",
	"height", // used for point-type paragraph styles
	"letter-spacing",
	"opacity",
	"padding-top",
	"padding-bottom",
	"text-align",
	"text-transform",
	"mix-blend-mode",
	"vertical-align" // for superscript
]

// These are base settings that are overridden by text block settings in
// the .ai document and settings contained in ai2html-config.json files
const defaultSettings: ai2HTMLSettings = {
	namespace: "g-",
	settings_version: "",
	create_promo_image: false,
	promo_image_width: 1024,
	image_format: ["auto"],
	write_image_files: true,
	responsiveness: "fixed",
	text_responsiveness: "dynamic",
	max_width: 0,
	output: "one-file",
	project_name: "",
	html_output_path: "ai2html-output/",
	html_output_extension: ".html",
	image_output_path: "ai2html-output/",
	image_source_path: "",
	image_alt_text: "",
	cache_bust_token: 0,
	create_config_file: false,
	create_settings_block: true,
	config_file_path: "",
	local_preview_template: "",
	png_transparent: false,
	png_number_of_colors: 128,
	jpg_quality: 60,
	center_html_output: true,
	use_2x_images_if_possible: true,
	use_lazy_loader: true,
	include_resizer_widths: true,
	include_resizer_css: true,
	include_resizer_script: false,
	inline_svg: false,
	svg_id_prefix: "",
	svg_embed_images: false,
	render_text_as: "html",
	render_rotated_skewed_text_as: "html",
	testing_mode: false,
	show_completion_dialog_box: true,
	clickable_link: "",
	last_updated_text: "",
	headline: "",
	leadin: "",
	summary: "",
	notes: "",
	sources: "",
	credit: "",

	settings_block: [
		"settings_version",
		"image_format",
		"responsiveness",
		"include_resizer_css",
		"output",
		"html_output_path",
		"image_output_path",
		"image_source_path",
		"local_preview_template",
		"png_number_of_colors",
		"jpg_quality",
		"headline",
		"leadin",
		"notes",
		"sources",
		"credit"
	],

	// list of settings to include in the config.yml file
	config_file: ["headline", "leadin", "summary", "notes", "sources", "credit"]
}
export {
	defaultFonts,
	basicCharacterReplacements,
	extraCharacterReplacements,
	caps,
	align,
	blendModes,
	defaultSettings,
	cssTextStyleProperties
}
