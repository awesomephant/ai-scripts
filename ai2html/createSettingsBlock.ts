import { forEach } from "../common/arrayUtils"
import { getAllArtboardBounds } from "../common/getAllArtboardBounds"
import makeRgbColor from "../common/makeRgbColor"
import { ai2HTMLSettings } from "./types"

export default function createSettingsBlock(
	settings: Partial<ai2HTMLSettings>,
	doc: Document,
	onsuccess: (message: string) => void
) {
	const bounds = getAllArtboardBounds(doc.artboards)
	const fontSize = 14
	const leading = 19
	const extraLines = 6
	const width = 400
	const left = bounds[0] - width - 50
	const top = bounds[1]
	const settingsLines: string[] = ["ai2html-settings"]
	var layer, rect, textArea, height

	forEach(settings.settings_block || [], (key: keyof ai2HTMLSettings) => {
		settingsLines.push(key + ": " + settings[key])
	})

	try {
		layer = doc.layers.getByName("ai2html-settings")
		layer.locked = false
	} catch (e) {
		layer = doc.layers.add()
		layer.zOrder(ZOrderMethod.BRINGTOFRONT)
		layer.name = "ai2html-settings"
	}

	height = leading * (settingsLines.length + extraLines)
	rect = layer.pathItems.rectangle(top, left, width, height)

	// @ts-expect-error bad upstream type def
	textArea = layer.textFrames.areaText(rect)
	textArea.textRange.autoLeading = false
	textArea.textRange.characterAttributes.leading = leading
	textArea.textRange.characterAttributes.fillColor = makeRgbColor([255, 255, 255])
	textArea.textRange.characterAttributes.size = fontSize
	textArea.contents = settingsLines.join("\n")
	textArea.name = "ai2html-settings"

	onsuccess("A settings text block was created to the left of all your artboards.")
	return textArea
}
