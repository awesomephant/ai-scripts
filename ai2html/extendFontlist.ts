import type { FontRule } from "./types"
import { forEach } from "../common/arrayUtils"
/**
 * Extends a FontRule[], overwriting if aifont matches, appending otherwise 
 */
export default function extendFontList(a: FontRule[], b: FontRule[]) {
    let index: Record<string, number> = {}
    let res = [...a]
    forEach(res, (f, i) => {
        index[f.aifont] = i
    })
    forEach(b, (f) => {
        if (f.aifont in index) {
            res[index[f.aifont]] = f // replace
        } else {
            res.push(f) // add
        }
    })
    return res
}