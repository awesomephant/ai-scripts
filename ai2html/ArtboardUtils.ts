import { forEach, indexOf } from "../common/arrayUtils"
import { aiBoundsToRect, boundsIntersect } from "../common/geometryUtils"
import { makeKeyword } from "../common/stringUtils"
import cleanObjectName from "./cleanObjectName"
import parseObjectName from "./parseObjectName"
import { ai2HTMLSettings, ArtboardGroupForOutput, ArtboardInfo } from "./types"

// Calls cb for every artboard in doc except those with names starting w/ "-"
function forEachUsableArtboard(doc: Document, cb: (ab: Artboard, i: number) => any) {
	for (let i = 0; i < doc.artboards.length; i++) {
		let ab = doc.artboards[i]
		if (!/^-/.test(ab.name)) {
			cb(ab, i)
		}
	}
}

function clearMatrixShift(m: Matrix, app: Application) {
	return app.concatenateTranslationMatrix(m, -m.mValueTX, -m.mValueTY)
}

/**
 * Returns an artboard's responsiveness setting or the global
 * responsiveness setting contained in the settings parameter
 */
function getArtboardResponsiveness(ab: Artboard, settings: ai2HTMLSettings) {
	var opts = parseObjectName(ab.name)
	if (opts.dynamic) return "dynamic" // ab name has ":dynamic" appended
	if (opts.fixed) return "fixed" // ab name has ":fixed" appended
	return settings.responsiveness //Default to document's responsiveness setting
}

// TODO: prevent duplicate names? or treat duplicate names an an error condition?
// (artboard name is assumed to be unique in several places)
function getArtboardName(ab: Artboard) {
	return cleanObjectName(ab.name)
}

function getLayerName(layer: Layer) {
	return cleanObjectName(layer.name)
}

function makeDocumentSlug(rawName: string) {
	return makeKeyword(rawName.replace(/ +/g, "-"))
}

function getRawDocumentName(doc: Document) {
	return doc.name.replace(/(.+)\.[aieps]+$/, "$1")
}
function getGroupContainerId(namespace: string, groupName: string) {
	return namespace + groupName + "-box"
}
function findUsableArtboards(doc: Document) {
	var arr: Artboard[] = []
	forEachUsableArtboard(doc, (ab) => {
		arr.push(ab)
	})
	return arr
}

// Prevent duplicate artboard names by appending width
// (Assumes dupes have different widths and have been named to form a group)
function getArtboardUniqueName(ab: Artboard, settings: ai2HTMLSettings, docSlug: string) {
	var suffix = ""
	if (settings.grouped_artboards) {
		suffix = "-" + Math.round(aiBoundsToRect(ab.artboardRect).width)
	}
	return getDocumentArtboardName(ab, docSlug) + suffix
}

/**
 * Return array of data records about each artboard, sorted from narrow to wide
 */
function getSortedArtboardInfo(artboards: Artboard[], settings: ai2HTMLSettings) {
	let res: ArtboardInfo[] = []
	forEach(artboards, function (ab) {
		res.push({
			effectiveWidth: getArtboardWidth(ab),
			responsiveness: getArtboardResponsiveness(ab, settings)
		})
	})
	res.sort((a, b) => {
		return a.effectiveWidth - b.effectiveWidth
	})
	return res
}

/**
 * Return the effective width of an artboard (the actual width, overridden by optional setting)
 */
function getArtboardWidth(ab: Artboard): number {
	var abSettings = parseObjectName(ab.name)
	return abSettings.width || aiBoundsToRect(ab.artboardRect).width
}

/**
 * Returns index of artboard with largest area (for promo image)
 */
function findLargestArtboardIndex(doc: Document) {
	let largestIdx = -1
	let largestArea = 0
	forEachUsableArtboard(doc, (ab, i) => {
		const info = aiBoundsToRect(ab.artboardRect)
		const area = info.width * info.height
		if (area > largestArea) {
			largestIdx = i
			largestArea = area
		}
	})
	return largestIdx
}

// Get [min, max] width range for the graphic (for optional config.yml output)
function getWidthRangeForConfig(settings: ai2HTMLSettings, doc: Document): [number, number] {
	const info = getSortedArtboardInfo(findUsableArtboards(doc), settings)
	const minAB = info[0]
	const maxAB = info[info.length - 1]
	let min, max
	if (!minAB || !maxAB) return [0, 0]
	min = settings.min_width || minAB.effectiveWidth
	if (maxAB.responsiveness == "dynamic") {
		max = settings.max_width || Math.max(maxAB.effectiveWidth, 1600)
	} else {
		max = maxAB.effectiveWidth
	}
	return [min, max]
}

function getDocumentArtboardName(ab: Artboard, docSlug: string) {
	return docSlug + "-" + getArtboardName(ab)
}

function calcProgressBarSteps(doc: Document): number {
	var n = 0
	forEachUsableArtboard(doc, () => {
		n += 2
	})
	return n
}

// create temp document (takes ~1.5s)
function makeTmpDocument(doc: Document, ab: Artboard) {
	const artboardBounds = ab.artboardRect
	let doc2 = app.documents.add(DocumentColorSpace.RGB, doc.width, doc.height, 1)
	doc2.pageOrigin = doc.pageOrigin // not sure if needed
	doc2.rulerOrigin = doc.rulerOrigin

	// The following caused MRAP:
	// doc2.artboards[0].artboardRect = ab.artboardRect;
	doc2.artboards[0].artboardRect = artboardBounds
	return doc2
}

function objectOverlapsArtboard(obj: PageItem, ab: Artboard) {
	return boundsIntersect(ab.artboardRect, obj.geometricBounds)
}

function objectOverlapsAnyArtboard(obj: PageItem, doc: Document) {
	var hit = false
	forEachUsableArtboard(doc, (ab) => {
		hit = hit || objectOverlapsArtboard(obj, ab)
	})
	return hit
}

/**
 * get range of container widths that an ab is visible as a [min,max] array
 * smallest artboard starts with 0, largest artboard ends with Infinity
 * values are inclusive and rounded
 * example: [0, 599]  [600, Infinity]
 */
function getArtboardVisibilityRange(
	ab: Artboard,
	group: ArtboardGroupForOutput,
	settings: ai2HTMLSettings
) {
	var thisWidth = getArtboardWidth(ab)
	let minWidth: number
	let nextWidth: number

	// find widths of smallest ab and next widest ab (if any)
	forEach(getSortedArtboardInfo(group.artboards, settings), function (info) {
		var w = info.effectiveWidth
		if (w > thisWidth && (!nextWidth || w < nextWidth)) {
			nextWidth = w
		}
		minWidth = Math.min(w, minWidth || Infinity)
	})
	return [thisWidth == minWidth ? 0 : thisWidth, !!nextWidth ? nextWidth - 1 : Infinity]
}

/**
 * Get range of widths that an ab can be sized
 */
function getArtboardWidthRange(
	ab: Artboard,
	group: ArtboardGroupForOutput,
	settings: ai2HTMLSettings
) {
	var responsiveness = getArtboardResponsiveness(ab, settings)
	var w = getArtboardWidth(ab)
	var visibleRange = getArtboardVisibilityRange(ab, group, settings)
	if (responsiveness == "fixed") {
		return [visibleRange[0] === 0 ? 0 : w, w]
	}
	return visibleRange
}

function findArtboardIndex(ab: Artboard, doc: Document) {
	return indexOf(doc.artboards, ab)
}

export {
	findUsableArtboards,
	forEachUsableArtboard,
	getArtboardName,
	getArtboardResponsiveness,
	getArtboardWidth,
	getGroupContainerId,
	getLayerName,
	getRawDocumentName,
	makeDocumentSlug,
	getSortedArtboardInfo,
	findLargestArtboardIndex,
	getWidthRangeForConfig,
	clearMatrixShift,
	makeTmpDocument,
	getDocumentArtboardName,
	calcProgressBarSteps,
	getArtboardUniqueName,
	objectOverlapsArtboard,
	objectOverlapsAnyArtboard,
	getArtboardVisibilityRange,
	getArtboardWidthRange,
	findArtboardIndex
}
