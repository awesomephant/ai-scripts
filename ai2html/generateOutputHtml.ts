// Wrap content HTML in a <div>, add styles and resizer script, write to a file

import { getGroupContainerId } from "./ArtboardUtils"
import { ai2HTMLSettings, ArtboardGroupForOutput, outputData } from "./types"
import { isTrue } from "../common/booleanUtils"
import makeResizerScript from "./makeResizerScript"
import getDateTimestamp from "../common/getDateTimestamp"
import { encodeHtmlEntities } from "../common/stringUtils"
import generatePageCss from "./generatePageCss"
import applyTemplate from "../common/applyTemplate"

export default function generateOutputHtml(
	content: outputData,
	group: ArtboardGroupForOutput,
	settings: ai2HTMLSettings,
	nameSpace: string,
	doc: Document,
	scriptVersion: string,
	pageName: string
) {
	const linkSrc = settings.clickable_link || ""
	let responsiveJs = ""
	const containerId = getGroupContainerId(nameSpace, pageName)
	const altTextId = containerId + "-img-desc"
	var textForFile, html, js, css, commentBlock
	var containerClasses: string[] = ["ai2html"]

	// accessibility features
	// prettier-ignore
	const ariaAttrs = `${settings.aria_role ? `role="${settings.aria_role}"` : ""} ${settings.alt_text ? ` aria-describedby="${altTextId}"` : ""}`

	if (
		isTrue(settings.include_resizer_script) &&
		// resizer CSS overrides the script setting
		!isTrue(settings.include_resizer_css)
	) {
		responsiveJs = makeResizerScript(containerId, nameSpace)
		containerClasses.push("ai2html-responsive")
	}

	// prettier-ignore
	commentBlock = `<!-- Generated by ai2html v${scriptVersion} - ${getDateTimestamp()} -->
<!-- ai file: ${doc.name} -->${settings.preview_slug ? `\n<!-- preview: ${settings.preview_slug} -->` : ""}
${settings.scoop_slug_from_config_yml ? `\n<!-- scoop: ${settings.scoop_slug_from_config_yml} -->` : ""}`

	// HTML
	html = `<div id="${containerId}" class="${containerClasses}" ${ariaAttrs}>`

	if (settings.alt_text) {
		html +=
			'<div class="' +
			nameSpace +
			'aiAltText" id="' +
			altTextId +
			'">' +
			encodeHtmlEntities(settings.alt_text) +
			"</div>"
	}
	if (linkSrc) {
		// optional link around content
		html += '\t<a class="' + nameSpace + 'ai2htmlLink" href="' + linkSrc + '">\r'
	}
	html += content.html
	if (linkSrc) {
		html += "\t</a>\r"
	}
	html += "</div>\r"

	// CSS
	css = `<style media="screen,print">
${generatePageCss(containerId, group, settings, nameSpace)}
${content.css}</style>`

	textForFile = `${commentBlock}${css}${html}${content.js}${responsiveJs}
<!-- End ai2html ${getDateTimestamp()} - ${getDateTimestamp()} -->`

	textForFile = applyTemplate(textForFile, settings)
	return textForFile
}
