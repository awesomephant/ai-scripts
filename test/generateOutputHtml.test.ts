import { it, expect, describe, vi } from "vitest"
import "html-validate/vitest"

import generateOutputHtml from "../ai2html/generateOutputHtml"
import { ai2HTMLSettings, outputData } from "../ai2html/types"

import _Document from "./__mocks__/Document"
import _Artboard from "./__mocks__/Artboard"
import groupArtboardsForOutput from "../ai2html/groupArtboardForOutput"

vi.useFakeTimers()

describe("generateOutputHtml", () => {
	vi.setSystemTime(new Date("2025-01-02 12:34"))

	const doc = new _Document()
	doc.name = "test.ai"
	doc.artboards = [
		new _Artboard("Artboard 1", [0, 0, 100, 100]), // 100x100
		new _Artboard("Artboard 2", [200, 0, 450, 150]) // 250x150
	]

	const settings = {
		scriptVersion: "123.45.56",
		namespace: "should not be included",
		alt_text: "alt text goes here",
		credit: "Alice",
		dark_mode_compatible: true,
		config_file: ["alt_text", "credit"],
		output: "one-file"
	} as ai2HTMLSettings

	const outputData: outputData = {
		html: "",
		css: "",
		js: ""
	}

	const groups = groupArtboardsForOutput(settings, "slug", doc as Document)

	const pageName = groups[0].groupName
	it("produces valid HTML", () => {
		const html = generateOutputHtml(
			outputData,
			groups[0],
			settings,
			"ns-",
			doc as Document,
			pageName
		)
		expect(html).toMatchSnapshot()
		expect(html).toHTMLValidate()
	})
})

