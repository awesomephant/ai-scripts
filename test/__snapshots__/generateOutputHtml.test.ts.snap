// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateOutputHtml > produces valid HTML 1`] = `
"<!-- Generated by ai2html v123.45.56 - 2025-01-02 12:34 -->
<!-- ai file: test.ai -->
<style media="screen,print">
#ns-test-box .ns-aiAltText {
	position:absolute;
	left:-10000px;
	width:1px;
	height:1px;
	overflow:hidden;
	white-space:nowrap;
}
#ns-test-box .ns-ai2htmlLink {
	display:block;
}
#ns-test-box p {
	margin:0;
}
#ns-test-box .ns-aiAbs {
	position:absolute;
}
#ns-test-box .ns-aiImg {
	position:absolute;
	top:0;
	display:block;
	width:100% !important;
}
#ns-test-box .ns-aiSymbol {
	position:absolute;
	box-sizing:border-box;
}
#ns-test-box .ns-aiPointText p {
	white-space:nowrap;
}
</style><div id="ns-test-box" class="ai2html"   aria-describedby="ns-test-box-img-desc"><div class="ns-aiAltText" id="ns-test-box-img-desc">alt text goes here</div></div>
<!-- End ai2html 2025-01-02 12:34 - 2025-01-02 12:34 -->"
`;
